### PR TITLE
AC_Avoid: move check for zero-desired-velocity into adjust_velocity_p…

### PR DESCRIPTION
**Pull request by peterbarker**
…olygon

This is the method which divides by the length of this vector; the
check belongs in here.

### CODE REVIEW

**Comment by rmackay9** [comment]
as @khancyr says it's a bit wasteful to wait this long before returning..  I don't mind I guess ..

**Comment by peterbarker** [comment]
"We ran out of CPU when the vehicle started to move" ;-)

**Comment by peterbarker** [comment]
I've added some more patches onto this (somewhat accidentally...).

Primarily adding tests, but also changing the pair-wise step of edges to be like we now do in Polygon_outside.

The tests added check velocity-based avoidance for a polygon fence, an `AP_Proximity_SITL` boundary (created from the same fence used in the polygon fence test), and for a SITL beacon fence at a 45-degree yaw.

File: libraries/AC_Avoidance/AC_Avoid.cpp [hunk 1]
```diff
@@ -474,8 +464,11 @@ void AC_Avoid::adjust_velocity_polygon(float kP, float accel_cmss, Vector2f &des
     const float speed = safe_vel.length();
     const Vector2f stopping_point_plus_margin = position_xy + safe_vel*((2.0f + margin_cm + get_stopping_distance(kP, accel_cmss, speed))/speed);
 
-    uint16_t i, j;
-    for (i = 0, j = num_points-1; i < num_points; j = i++) {
+    for (uint16_t i=0; i<num_points; i++) {
```
Comment (WickedShell) [inline libraries/AC_Avoidance/AC_Avoid.cpp]: This is a pretty debatable exchange. It's technically 1 less instruction, but it introduces an extra branch which has a larger runtime hit then a single 1 cycle instruction. It's also much wordier like this, but I'd guess that some would find it more readable?

At a minimum please bring in spacing around the `i < num_points`.

Comment (rmackay9) [reply libraries/AC_Avoidance/AC_Avoid.cpp]: I find it more readable.. re the spacing, we're a bit all over the place for for loops..  i was also recently wondering (in a different part of the code) if we should have spaces in these situations or not..

Comment (WickedShell) [reply libraries/AC_Avoidance/AC_Avoid.cpp]: Definitely spaces for readability! I'd argue pretty forcefully for that one.

**Comment by peterbarker** [comment]
On Wed, 29 May 2019, WickedShell wrote:

> +    for (uint16_t i=0; i<num_points; i++) {
> Definitely spaces for readability! I'd argue pretty forcefully for that one.

I find the spacing I've used here much more readable.  Obviously.

**Comment by OXINARF** [comment]
+200000 for spaces around operators! Anyone that doesn't add them should be banished from coding 😄

**Comment by peterbarker** [comment]
On Wed, 29 May 2019, Francisco Ferreira wrote:

> +200000 for spaces around operators! Anyone that doesn't add them should be banished from coding 😄

We're going to run short on coders in that case.

-*- mode: grep; default-directory: "~/rc/ardupilot/ArduPlane/" -*-
Grep started at Thu May 30 10:15:11

git --no-pager grep -n -e for\ \( -- *
Parameters.cpp:1460:    for (uint8_t i=0; i<NUM_SERVO_CHANNELS; i++) {
geofence.cpp:125:    for (i=0; i<g.fence_total; i++) {
quadplane.cpp:518:    for (uint8_t i=0; i<num_motors; i++) {
quadplane.cpp:686:    for (uint8_t i=0; i<8; i++) {
tailsitter.cpp:329:    for (uint8_t i=0; i<ARRAY_SIZE(functions); i++) {
tiltrotor.cpp:244:    for (uint8_t i=0; i<num_motors; i++) {
tiltrotor.cpp:255:    for (uint8_t i=0; i<num_motors; i++) {
tiltrotor.cpp:269:        for (uint8_t i=0; i<num_motors; i++) {
tiltrotor.cpp:295:    for (uint8_t i=0; i<num_motors; i++) {
tiltrotor.cpp:305:    for (uint8_t i=0; i<num_motors; i++) {

Grep finished (matches found) at Thu May 30 10:15:11


-*- mode: grep; default-directory: "~/rc/ardupilot/ArduCopter/" -*-
Grep started at Thu May 30 10:16:42

git --no-pager grep -n -e for\ \( -- *
ArduCopter.cpp:464:    for (uint8_t i=0; i<gps.num_sensors(); i++) {
GCS_Mavlink.cpp:197:    for (uint8_t i=0; i<ARRAY_SIZE(axes); i++) {
Parameters.cpp:1155:    for (uint8_t i=0; i<table_size; i++) {
Parameters.cpp:1160:    for (uint8_t i=0; i<table_size; i++) {
Parameters.cpp:1165:    for (uint8_t i=0; i<table_size; i++) {
Parameters.cpp:1170:    for (uint8_t i=0; i<table_size; i++) {
Parameters.cpp:1181:    for (uint8_t i=0; i<table_size; i++) {
Parameters.cpp:1318:        for (uint8_t i=0; i<table_size; i++) {
Parameters.cpp:1338:        for (uint8_t i=0; i<table_size; i++) {
compassmot.cpp:37:    for (uint8_t i=0; i<COMPASS_MAX_INSTANCES; i++) {
compassmot.cpp:52:    for (uint8_t i=0; i<compass.get_count(); i++) {
compassmot.cpp:113:    for (uint8_t i=0; i<compass.get_count(); i++) {
compassmot.cpp:122:    for (uint8_t i=0; i<compass.get_count(); i++) {
compassmot.cpp:169:            for (uint8_t i=0; i<compass.get_count(); 
i++) {
compassmot.cpp:175:            for (uint8_t i=0; i<compass.get_count(); 
i++) {
compassmot.cpp:182:                for (uint8_t i=0; 
i<compass.get_count(); i++) {
compassmot.cpp:191:                for (uint8_t i=0; 
i<compass.get_count(); i++) {
compassmot.cpp:205:                for (uint8_t i=0; 
i<compass.get_count(); i++) {
compassmot.cpp:210:                for (uint8_t i=0; 
i<compass.get_count(); i++) {
compassmot.cpp:240:        for (uint8_t i=0; i<compass.get_count(); i++) {
mode_autotune.cpp:136:    for (uint8_t i=0; i<ARRAY_SIZE(map); i++) {
mode_flowhold.cpp:180:        for (uint8_t i=0; i<2; i++) {
mode_flowhold.cpp:436:    for (uint8_t i=0; i<2; i++) {
motors.cpp:163:    for (uint8_t i=0; i<=10; i++) {
setup.cpp:17:    for (uint8_t i=0; i<compass.get_count(); i++) {
setup.cpp:39:        for (uint8_t i=0; i<compass.get_count(); i++) {
setup.cpp:61:    for (int i = 0; i < 40; i++) {
toy_mode.cpp:724:    for (uint8_t i=0; i<4; i++) {
toy_mode.cpp:742:    for (uint8_t i=0; i<4; i++) {
toy_mode.cpp:760:    for (uint8_t i=0; i<4; i++) {
toy_mode.cpp:769:    for (uint8_t i=0; i<4; i++) {
toy_mode.cpp:988:    for (uint8_t i=0; i<num_motors; i++) {
toy_mode.cpp:1012:        for (uint8_t i=0; i<4; i++) {
toy_mode.cpp:1024:        for (uint8_t i=0; i<4; i++) {
toy_mode.cpp:1035:        for (uint8_t i=0; i<4; i++) {
toy_mode.cpp:1042:    for (uint8_t i=0; i<4; i++) {

Grep finished (matches found) at Thu May 30 10:16:42

**Comment by OXINARF** [comment]
@peterbarker I'm perfectly aware of still many wrong cases. Fortunately we have been moving in the right direction 😜

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request merged**